fish-rust/src/tinyexpr.rs

```diff
@@ -0,0 +1,714 @@
+                    ' ' | '\t' | '\n' | '\r' => return Some((1, None)),
+                    '=' | '>' | '<' | '&' | '|' | '!' => {
+                        self.set_error(ErrorKind::LogicalOperator, None);
+                        Token::Error
+                    }
+                    _ => {
+                        self.set_error(ErrorKind::MissingOperator, None);
+                        Token::Error
+                    }
+                };
+
+                Some((1, Some(tok)))
+            }
+        }
+    }
+
+    fn next_token(&mut self) {
+        self.current = loop {
+            let Some((consumed, token)) = self.get_token() else {
+                break Token::End;
+            };
+
+            self.pos += consumed;
+            if let Some(token) = token {
+                break token;
+            }
+        };
+    }
+
+    /// ```
+    /// <base>   = <constant> |
+    ///            <function-0> {"(" ")"} |
+    ///            <function-1> <power> |
+    ///            <function-X> "(" <expr> {"," <expr>} ")" |
+    ///            "(" <list> ")"
+    /// ```
+    fn base(&mut self) -> f64 {
+        match self.current {
+            Token::Number(n) => {
+                let after_first = self.pos;
+
+                self.next_token();
+                if let Token::Number(_) | Token::Function(_) = self.current {
+                    // Two numbers after each other:
+                    // math '5 2'
+                    // math '3 pi'
+                    // (of course 3 pi could also be interpreted as 3 x pi)
+
+                    // The error should be given *between*
+                    // the last two tokens.
+                    let num_whitespace = self.start[after_first..]
+                        .chars()
+                        .take_while(|&c| " \t\n\r".contains(c))
+                        .count();
+
+                    self.set_error(
+                        ErrorKind::MissingOperator,
+                        Some((after_first, num_whitespace)),
+                    );
+                }
+
+                n
+            }
+            Token::Function(f) => {
+                self.next_token();
+                let have_open = matches!(self.current, Token::Open);
+                if have_open {
+                    // If we *have* an opening parenthesis,
+                    // we need to consume it and
+                    // expect a closing one.
+                    self.next_token();
+                }
+
+                if let Function::Fn0(f) = f {
+                    if have_open {
+                        if let Token::Close = self.current {
+                            self.next_token();
+                        } else if self.no_specific_error() {
+                            self.set_error(ErrorKind::MissingClosingParen, None);
+                        }
+                    }
+                    return f();
+                }
+                if let Function::Constant(n) = f {
```

Comment:
Why not use `arity == 0` for both Fn0 and Constants as before?

Reply:
It involves an (unreachable) panic because you still have to extract the enum variant's content inside the `if`. I guess that's still prettier than duplicating the `if` body? Please take a look.